containers/src/Data/IntMap/Internal.hs

```diff
@@ -598,7 +599,23 @@ notMember k m = not $ member k m
 lookup :: Key -> IntMap a -> Maybe a
 lookup !k = go
   where
-    go (Bin p m l r) | nomatch k p m = Nothing
+    go (Bin _p m l r)
+                     | zero k m  = go l
+                     | otherwise = go r
+    go (Tip kx x) | k == kx   = Just x
+                  | otherwise = Nothing
+    go Nil = Nothing
+
+-- | /O(min(n,W))/. 'query' has identical behaviour to 'Data.IntMap.Internal.lookup', but
+-- will fail faster in the case that the key is not present. It is otherwise
+-- preferable to use `Data.IntMap.Internal.lookup'.
+
+-- See Note: Local 'go' functions and capturing]
+query :: Key -> IntMap a -> Maybe a
+query !k = go
+  where
+    go (Bin p m l r)
+                     | nomatch k p m = Nothing
                      | zero k m  = go l
                      | otherwise = go r
```

Comment:
```suggestion
    go (Bin p m l r) | nomatch k p m = Nothing
                     | zero k m  = go l
                     | otherwise = go r
```
Same as above.